capsules/src/dals/README.md

```diff
@@ -0,0 +1,100 @@
+peripheral capsules consist of the **AppLoaderClient**, **Decompressor**, **Verifier**, and 
+**Validator** capsules.  Each capsule is named after the trait or interface it should implement.
+These interfaces are defined in `interfaces.rs`.
+
+The way these capsules are interconnected is diagrammed below.
+
+
+                                            AppLoaderClient
+                                                 |
+                                                 |
+                                                 |
+                                                 |
+                         Validator ---------  AppLoader  ----------  Decompressor
+                                                 |
+                                                 |
+                                                 |
+                                                 |
+                                              Verifier
+                                        
+
+There exists a circular reference between the **AppLoader** capsule and each of the peripheral
+capsules.  This enables the **AppLoader** to call functions implemented by a peripheral capsule and 
+vice versa.
+
+The peripheral modules are detailed below.  Please see `interfaces.rs` for more information on functions and such.
+
+
+## AppLoaderClient
+
+This capsule implements `pub trait AppLoaderClient<'a>`.  This capsule can be any capsule
+which receives data via standard communication protocols for embedded systems such as BLE, 
+Zigbee, USB.  This capsule's main function is to send this raw received data buffer-by-buffer
+to the **AppLoader** capsule.  In order to prevent bottleneck issues, this capsule must send an ack 
+only after its buffer is returned to it from the **AppLoader**.
+
+This capsule must also include a buffer the size of one data segment (different for different protocols) which
+it sends to the **AppLoader** to send one segment of the received data.
+
+## Decompressor
+
+**DALS** support data decompression methods which "operate on-the-fly".  That is, the received app data
+is decompressed segment-by-segment. This capsule must implement `pub trait Decompressor<'a>` and contain 
+a decompression algorithm which conforms to the above constraints.
+
+This capsule must include a buffer in which to store the resulting decompressed data which is passed
+back to the **AppLoader**.
+
+
+## Verifier
+
+This capsule implements `pub trait Verifier<'a>`.  This capsule contains an algorithm which can
+perform data verification on the app data loaded into flash.  Such an algorithm could be used to 
+verify that no bit-errors occured during wireless transmission or that the loaded app is not malware.
+Examples of this include checksums, digital signatures or more complex cryptographic computations.
+
+This capsule must contain an immutable reference to the app binary data in flash space.
+
+## Validator
```

Comment:
Crypto checks would probably occur here, to make sure the app matches signatures that we allow to run for example.

Reply:
@bradjc and I discussed we think its a good idea to leave crypto checks in Verification stage.  Validation stage is designed to be more based around system designer policies.  For example, the user/designer could institute a policy to not run apps which need x, y, or z hardware resources.  Verification is geared more towards running some algorithm on the whole app binary in flash which checks for bit-errors or digital signature matches.  Let me know if this works.

Reply:
Won't a signature require the entire app to be loaded though?

Reply:
Sorry for missing this.  The pipeline is loading --> decompressing --> verification --> validation.  After the decompressing phase the `AppLoader` will write the app binary into flash space.  The verification phase can then process on the app binary data in flash to perform any crypto checks/signature-checking.